apps/domain/test/domain/relays_test.exs

```diff
@@ -415,6 +416,54 @@ defmodule Domain.RelaysTest do
     end
   end
 
+  describe "list_connected_relays_for_resource/1" do
+    test "returns empty list when there are no online relays", %{account: account} do
+      resource = ResourcesFixtures.create_resource(account: account)
+
+      RelaysFixtures.create_relay(account: account)
+
+      RelaysFixtures.create_relay(account: account)
+      |> RelaysFixtures.delete_relay()
+
+      assert list_connected_relays_for_resource(resource) == {:ok, []}
+    end
+
+    test "returns list of connected relays", %{account: account} do
+      resource = ResourcesFixtures.create_resource(account: account)
+      relay = RelaysFixtures.create_relay(account: account)
+      stamp_secret = Ecto.UUID.generate()
+
+      assert connect_relay(relay, stamp_secret) == :ok
+
+      assert {:ok, [connected_relay]} = list_connected_relays_for_resource(resource)
+
+      assert connected_relay.id == relay.id
+      assert connected_relay.stamp_secret == stamp_secret
+    end
+  end
+
+  describe "generate_username_and_password/1" do
+    test "returns username and password", %{account: account} do
+      relay = RelaysFixtures.create_relay(account: account)
+      stamp_secret = Ecto.UUID.generate()
+      relay = %{relay | stamp_secret: stamp_secret}
+      expires_at = DateTime.utc_now() |> DateTime.add(3, :second)
+
+      assert %{username: username, password: password, expires_at: expires_at_unix} =
+               generate_username_and_password(relay, expires_at)
+
+      assert [username_expires_at_unix, username_salt] = String.split(username, ":", parts: 2)
+      assert username_expires_at_unix == to_string(expires_at_unix)
+      assert DateTime.from_unix!(expires_at_unix) == DateTime.truncate(expires_at, :second)
+
+      expected_hash =
+        :crypto.hash(:sha256, "#{expires_at_unix}:#{stamp_secret}:#{username_salt}")
+        |> Base.encode64(padding: false, case: :lower)
```

Comment:
@conectado @thomaseizinger @jamilbk This is a test that showcases how hashing is performed and how the hash is encored, which will be needed for comparison on the Rust end. 

We can change the algo but Sha256 is considered secure, Blake3 is not supported natively by Erlang so installing nifs just to use it is an overkill that won't improve security (only hashing performance).

Reply:
It would be great to have some language-agnostic test vectors, then I can just use them in my tests as well. I'll make some and add them to the spec.

Reply:
Made a proposal here: https://github.com/firezone/product/pull/599